node/libs/crypto/src/bn254/mod.rs

```diff
@@ -0,0 +1,239 @@
+//! BLS signature scheme for the BN254 curve.
+
+use std::collections::BTreeMap;
+use std::fmt::Debug;
+use std::hash::Hasher;
+
+use anyhow::anyhow;
+use ark_bn254::{Bn254, Fr, G1Projective as G1, G2Projective as G2};
+use ark_ec::pairing::Pairing as _;
+use ark_ec::pairing::PairingOutput;
+use ark_ec::Group as _;
+use ark_serialize::{CanonicalDeserialize, CanonicalSerialize};
+use num_traits::Zero as _;
+
+pub use error::Error;
+
+use crate::ByteFmt;
+
+#[doc(hidden)]
+pub mod error;
+
+#[cfg(test)]
+mod tests;
+
+pub mod hash;
+mod testonly;
+
+/// Type safety wrapper around a scalar value.
+pub struct SecretKey(pub Fr);
+
+impl SecretKey {
+    /// Gets the corresponding [`PublicKey`] for this [`SecretKey`]
+    pub fn public(&self) -> PublicKey {
+        let p = G2::generator() * self.0;
+        PublicKey(p)
+    }
+
+    /// Produces a signature using this [`SecretKey`]
+    pub fn sign(&self, msg: &[u8]) -> Signature {
+        let hash_point = hash::hash_to_g1(msg);
+        let sig = hash_point * self.0;
+        Signature(sig)
+    }
+}
+
+impl ByteFmt for SecretKey {
+    fn decode(bytes: &[u8]) -> anyhow::Result<Self> {
+        Fr::deserialize_compressed(bytes)
+            .map(Self)
+            .map_err(|e| anyhow!("failed to decode secret key: {e:?}"))
+    }
+
+    fn encode(&self) -> Vec<u8> {
+        let mut buf = Vec::new();
+        self.0.serialize_compressed(&mut buf).unwrap();
+        buf
+    }
+}
+
+/// Type safety wrapper around G2.
+#[derive(Clone)]
+pub struct PublicKey(pub G2);
+
+impl PartialEq for PublicKey {
+    fn eq(&self, other: &Self) -> bool {
+        ByteFmt::encode(self).eq(&ByteFmt::encode(other))
+    }
+}
+
+impl Eq for PublicKey {}
```

Comment:
This should not be necessary. I think you are able to just derive this trait, since the G2 type already implements it.

Reply:
Fixed.